vec/src/main.rs

```diff
@@ -0,0 +1,365 @@
+
+                        i.insert(0, 0);
+                        let diffs = i.windows(2).map(|w| w[1] - w[0]);
+
+                        let mut last = iter.nth(0).expect("exhausted iterator");
+                        for (i, diff) in diffs.enumerate() {
+                            if diff > 0 {
+                                last = iter.nth(diff - 1).expect("exhausted iterator");
+                            }
+                            indices[order[i]] = last;
+                        }
+
+                        // and finally we convert our new indices into an iterator
+                        iter = Box::new(indices.into_iter())
+                    }
+                },
+                Subset::Range(range) => {
+                    iter = Box::new(
+                        iter.skip(range.start)
+                            .enumerate()
+                            .take_while(move |(i, _)| i < &(range.end - range.start))
+                            .map(|(_, v)| v)
+                    ) as Self::IntoIter
+                },
+            }
+        }
+
+        iter
+    }
+}
+
+impl<T: Atomic> From<Vec<T>> for Vector<T> {
+    fn from(value: Vec<T>) -> Self {
+        Vector::Subset(Rc::new(RefCell::new(value)), Subsets(Vec::new()))
+    }
+}
+
+impl From<usize> for Subset {
+    fn from(value: usize) -> Self {
+        Subset::Indices(vec![value])
+    }
+}
+
+impl From<Range<usize>> for Subset {
+    fn from(value: Range<usize>) -> Self {
+        Subset::Range(value)
+    }
+}
+
+impl From<Vec<usize>> for Subset {
+    fn from(value: Vec<usize>) -> Self {
+        Subset::Indices(value)
+    }
+}
+
+fn main() {
+    // print out equivalent R code to mock up an interface
+    println!("R 0.v.0 (fake R)");
+
+    let x: Vector<_> = (1..=10).into_iter().collect::<Vec<_>>().into();
+    let y: Vector<_> = vec![10, 9].into();
+    println!("> x <- {}", x);
+    println!("> y <- {}\n> ", y);
+
+    // subset x by some indices
+    let x_subset_1 = x.subset(2..8).subset(vec![1, 2, 4]);
+    println!("> # range and (ordered) vector indexing\n> x[3:8][c(2, 3, 5)]\n{}\n> ", x_subset_1);
+
+    // subset x by some indices
+    let x_subset = x.subset(2..8).subset(vec![4, 2, 2]);
+    println!("> # out-of-order indices\n> x[3:8][c(5, 3, 3)]\n{}\n> ", x_subset);
+
+    // modify
+    x_subset.assign(y);
+    println!("> x[3:8][c(5, 3)] <- y");
+    println!("> x\n{}", x);
+}
+
+#[cfg(test)]
+ mod test {
+    use super::*;
+
+    #[test]
+    fn subset_range() {
+        let x: Vector<_> = (1..=10).into_iter().collect::<Vec<_>>().into();
+        let Vector::Subset(res, _) = x.subset(2..6).materialize();
+        assert_eq!(res.clone().borrow().to_owned(), vec![3, 4, 5, 6])
+    }
+
+    #[test]
+    fn subset_sequential_indices() {
+        let x: Vector<_> = (1..=10).into_iter().collect::<Vec<_>>().into();
+        let Vector::Subset(res, _) = x.subset(vec![2, 3, 4, 5]).materialize();
```

Comment:
What would be the use case where you _wouldn't_ materialize the results? Would it be possible to modify in place?

Reply:
The primary use case is for assignment. If you're doing `x[1:1e6][4:10][[2]] <- 4`, then we'd want to subset x a few times for the various `[[<-` calls, then assign to the index explicitly. We need a mutable reference that we can pass around so that we aren't materializing new vectors with each subset level.

I _think_ R currently does this naive approach, so you can hit some pretty nasty performance issues with deeply nested indexing.